undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterRefDefinition.java

```diff
@@ -128,15 +131,44 @@ protected void performRuntime(OperationContext context, ModelNode operation, Mod
                 predicate = PredicateParser.parse(predicateString, getClass().getClassLoader());
             }
 
-            int priority = PRIORITY.resolveModelAttribute(context, operation).asInt();
-            final ServiceTarget target = context.getServiceTarget();
+            final int priority = PRIORITY.resolveModelAttribute(context, operation).asInt();
+            final CapabilityServiceTarget target = context.getCapabilityServiceTarget();
             final ServiceName sn = UndertowService.getFilterRefServiceName(address, name);
-            final ServiceBuilder<?> sb = target.addService(sn);
-            final Consumer<UndertowFilter> frConsumer = sb.provides(sn);
-            final Supplier<HandlerWrapper> fSupplier = sb.requires(UndertowService.FILTER.append(name));
+            final CapabilityServiceBuilder<?> sb = target.addCapability(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition());
+            final Consumer<UndertowFilter> frConsumer = sb.provides(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition() , sn);
+            final Supplier<HandlerWrapper> fSupplier = sb.requiresCapability(Capabilities.CAPABILITY_FILTER, HandlerWrapper.class, name);
             final Supplier<FilterLocation> lSupplier = sb.requires(locationSN);
+            sb.setInitialMode(Mode.ACTIVE);
             sb.setInstance(new FilterService(frConsumer, fSupplier, lSupplier, predicate, priority));
             sb.install();
         }
+
+        @Override
+        protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource)
+                throws OperationFailedException {
+            super.recordCapabilitiesAndRequirements(context, operation, resource);
+            //NOTE: its dynamic by design.
+            context.registerCapability(FilterCapabilities.FILTER_REF_CAPABILITY.getDefinition().fromBaseCapability(context.getCurrentAddress()));
+        }
```

Comment:
You do not need to override this method. Capability registration is already handled via the super implementation.
See: https://github.com/wildfly/wildfly-core/blob/main/controller/src/main/java/org/jboss/as/controller/AbstractAddStepHandler.java#L216